src/target/riscv/riscv.c

```diff
@@ -3790,6 +3791,22 @@ int riscv_openocd_poll(struct target *target)
 		LOG_TARGET_DEBUG(target, "resume all");
 		riscv_resume(target, true, 0, 0, 0, false);
 	} else if (halted && running) {
+		foreach_smp_target(entry, targets)
+		{
+			struct target *t = entry->target;
+			if (t->state == TARGET_HALTED) {
+				riscv_reg_t dcsr;
+				if (riscv_reg_get(t, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
+					return ERROR_FAIL;
+				if (get_field(dcsr, CSR_DCSR_CAUSE) == CSR_DCSR_CAUSE_GROUP)
+					cause_groups++;
+				else
+					break;
+			}
+		}
+		if (halted == cause_groups)
+			return ERROR_OK;
```

Comment:
(1) I would recommend to keep a counter of how many times this situation has occurred. If it reaches a certain limit, OpenOCD should:
- stop waiting for the harts to halt, 
- print an error to the user, and 
- try to halt all the harts directly (riscv_halt()).

(2) Instead of returning from `riscv_openocd_poll()` and waiting for the next poll interval, it would perhaps be better to create a loop inside this function. This would allow to immediately re-poll the hart state and react quickly. 
 - @aap-sc @zqb-all What is your opinion on that? 
 - If you are in favor of it but don't feel like implementing it in this PR, consider adding a TODO comment.

Reply:
> create a loop inside this function. This would allow to immediately re-poll the hart state and react quickly.

Sounds good, this helps smp harts finish state adjustment faster.